### PR TITLE
Handle "too old" error when re-establishing Kubernetes watches

### PR DESCRIPTION
**Pull request by obeattie**
When a watch against the k8s API is terminated (because of a time-out, connection failure, etc), it must be re-established. Kubernetes only allows watching from a "recent" version, so when re-establishing the watch, it's possible it will be rejected with a "resource too old" error. In this case, it's necessary to refresh the resource list to verify no updates were missed, and then re-establish the watch from the newly-received version number.

🔋  Apologies again this one doesn't include tests. I will write some, but I'll do that as a separate PR in a few days if that's okay.

### CODE REVIEW

**Comment by wmorgan** [comment]
Nice corner case. Thanks, Oliver!

On Monday, June 13, 2016, Oliver Beattie <notifications@github.com> wrote:

> When a watch against the k8s API is terminated (because of a time-out,
> connection failure, etc), it must be re-established. Kubernetes only allows
> watching from a "recent" version, so when re-establishing the watch, it's
> possible it will be rejected with a "resource too old" error. In this case,
> it's necessary to refresh the resource list to verify no updates were
> missed, and then re-establish the watch from the newly-received version
> number.
>
> 🔋 Apologies again this one doesn't include tests. I will write some, but
> I'll do that as a separate PR in a few days if that's okay.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/BuoyantIO/linkerd/pull/473
> Commit Summary
>
>    - Handle "too old" error when re-establishing Kubernetes watches
>
> File Changes
>
>    - *M* k8s/src/main/scala/io/buoyant/k8s/Watchable.scala
>    <https://github.com/BuoyantIO/linkerd/pull/473/files#diff-0> (16)
>
> Patch Links:
>
>    - https://github.com/BuoyantIO/linkerd/pull/473.patch
>    - https://github.com/BuoyantIO/linkerd/pull/473.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BuoyantIO/linkerd/pull/473>, or mute the thread
> <https://github.com/notifications/unsubscribe/AAAPutvvl96X-spF_nCkXvKIhGoznxtGks5qLU_fgaJpZM4I0SXS>
> .
>


-- 
-William

**Comment by olix0r** [comment]
⭐  thanks for doing this, @obeattie!

**Comment by adleong** [comment]
Thanks for the fix, @obeattie!  I think it's a good idea to test this corner case.  Would you like to add tests to this PR in a few days when you have time?  Otherwise, one of us can take it over and add the tests.  

Thanks again!

**Comment by obeattie** [comment]
Sure, I'm more than happy to add the tests to this PR in a few days :)

**Comment by obeattie** [comment]
I've added a test for this. Apologies about the amount of copy + paste, feel free to refactor the test if you like.

File: k8s/src/test/scala/io/buoyant/k8s/v1/ApiTest.scala [hunk 1]
```diff
@@ -254,4 +254,68 @@ class ApiTest extends FunSuite with Awaits with Exceptions {
...
+    var finalResponse: Response = null
+    val service = Service.mk[Request, Response] { req =>
+      reqCount += 1
+      reqCount match {
+        case 1 =>
+          assert(req.uri == s"/api/v1/endpoints?watch=true&resourceVersion=$ver")
+          val rsp = Response()
+          rsp.version = req.version
+          rsp.status = Status.Gone
+          val msg = Buf.Utf8("""{"type":"ERROR","object":{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"401: The event in requested index is outdated and cleared (the requested history has been cleared [4770862/4659254]) [4771861]"}}""")
```
Comment (adleong) [inline k8s/src/test/scala/io/buoyant/k8s/v1/ApiTest.scala]: You can do `rsp.contentString =`

File: k8s/src/test/scala/io/buoyant/k8s/v1/ApiTest.scala [hunk 2]
```diff
@@ -254,4 +254,68 @@ class ApiTest extends FunSuite with Awaits with Exceptions {
       }
     }
   }
+
+  test("watch too old") {
+    val ver = "4659253"
+    var reqCount = 0
+    var finalResponse: Response = null
```
Comment (adleong) [inline k8s/src/test/scala/io/buoyant/k8s/v1/ApiTest.scala]: I think this is unused

**Comment by adleong** [comment]
Thanks, @obeattie, this is great!  Sorry for the delay getting this merged.
